File: src/main/java/com/d3bot/events/utilities/UrlFetcher.java
package com.d3bot.events.utilities;

import org.springframework.stereotype.Component;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

@Component
public class UrlFetcher {

    private final HttpClient httpClient;

    public UrlFetcher(HttpClient httpClient) {
        this.httpClient = httpClient;
    }

    public String fetch(String url) throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .GET()
                .build();
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() != 200) {
            throw new IOException("Request failed with status " + response.statusCode() + ": " + response.body());
        }
        return response.body();
    }
}

File: internal/utilities/url_fetcher.go
package utilities

import (
	"fmt"
	"io"
	"net/http"
)

type URLFetcher struct {
	client *http.Client
}

func NewURLFetcher(client *http.Client) *URLFetcher {
	return &URLFetcher{client: client}
}

func (f *URLFetcher) Fetch(url string) (string, error) {
	resp, err := f.client.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(body))
	}
	return string(body), nil
}
